### ISSUE TITLE
Modifying capsule state from another capsule listener failing

### ISSUE THREAD

**Issue by busslina** [GregoryConrad/rearch-dart#97]
Let say I have two capsules based on `use.state`: A and B.

In a React component mount function I'm listening individually both A and B.

In B listener I'm modifying A state (by CapsuleContainer.read), but A listener is not getting executed.

A listener is not failing, because when modified in other parts is getting triggered. It's B listener which is caching A capsule, so it's not calling a fresh A setState. Is the first time I try to modify a Capsule from another Capsule listener.

Any idea on what's happening?

**Comment by busslina** [comment]
In B listener I execute conditionally the A setState, but I store the setter in a variable before the if/else

**Comment by GregoryConrad** [comment]
I'm having a bit of a hard time following from your description; could you please provide a minimal code sample?

**Comment by busslina** [comment]
Everything happens on `componentDidMount` method. The `A` listener is `_cancelLoadingModeListener` and `B` listener is `_cancelWebsocketStateListener`. I tried multiple workarounds, but none with success.

```dart
import 'package:busslina_dart_web_lib/lib.dart' as wlib;
import 'package:pandemic_plandemic_web/lib.dart';

import 'package:react/react.dart';

var bodyApp = registerComponent2(() => _BodyApp());

/// Initiated on loading state.
class _BodyApp extends Component2 with wlib.ComponentMixin {
  static const _loadingField = 'loading';
  static const _softLoadingField = 'soft-loading';

  late final void Function() _cancelLoadingModeListener;
  late final void Function() _cancelSoftLoadingModeListener;
  late final void Function() _cancelRouterListener;
  late final void Function() _cancelWebsocketStateListener;

  bool get _loading => state[_loadingField];

  @override
  Map get initialState => {
        ...super.initialState,
        _loadingField: cc.read(wlib.loadingModeCapsule).loading,
        _softLoadingField: cc.read(wlib.softLoadingModeCapsule).loading,
      };

  @override
  render() {
    final loading = _loading;
    final softLoading = state[_softLoadingField];

    return div({
      ...processedProperties,
    }, [
      // Header
      if (!loading)
        headerWidget({
          'key': 'header',
        }),

      // Current route
      if (!loading)
        _currentRoute?.component({
          'key': 'current-route',
          wlib.softLoadingField: softLoading,
        }),

      // Footer
      if (!loading)
        footerWidget({
          'key': 'footer',
        }),

      // Loading
      if (loading)
        _loadingSpinner({
          'key': 'loading-spinner',
        }),

      // Soft loading
      if (softLoading)
        _softLoadingSpinner({
          'key': 'soft-loading-spinner',
        }),
    ]);
  }

  @override
  void componentDidMount() {
    super.componentDidMount();

    debug('_BodyApp.componentDidMount()');

    // final log = cc.read(wlib.loggerCapsule);

    // Listening to loading mode
    _cancelLoadingModeListener = cc.listen((use) {
      final loading = use(wlib.loadingModeCapsule).loading;

      debug('LOADING: $loading');

      setState({_loadingField: loading});
    });

    // Listening to soft loading mode
    _cancelSoftLoadingModeListener = cc.listen((use) {
      setState({_softLoadingField: use(wlib.softLoadingModeCapsule).loading});
    });

    // Listening to router
    _cancelRouterListener = cc.listen((use) {
      final routerController = use(wlib.routerCapsule);

      if (!routerController.initied) {
        return;
      }

      forceUpdate();
    });

    // cc.read(appCapsule).connectToWebsocketServer();

    // delay(const Duration(seconds: 10)).then((value) {
    //   cc.read(wlib.loadingModeCapsule).on();
    // });

    // Listening to websocket state
    _cancelWebsocketStateListener = cc.listen((use) {
      final state = use(wlib.websocketStateCapsule);

      final loadingMode = cc.read(wlib.loadingModeCapsule);
      // final loadingMode = use(wlib.loadingModeCapsule);

      debug('websocket state listener -- $state');

      // TODO: min delay between attempts (5 seconds)

      if (state.disconnected) {
        debug('SETTING LOADING ON -- current: ${loadingMode.loading}');

        // Setting loading
        loadingMode.on();

        // await delay(10.seconds);
        delay(10.seconds).then((value) {
          // Connecting to websocket server
          cc.read(appCapsule).connectToWebsocketServer();
        });

        // Connecting to websocket server
        // app.connectToWebsocketServer();
      }
    });
  }

  @override
  void componentWillUnmount() {
    debug('_BodyApp.componentWillUnmount()');

    // Rearch listeners
    _cancelLoadingModeListener();
    _cancelSoftLoadingModeListener();
    _cancelRouterListener();
    _cancelWebsocketStateListener();

    super.componentWillUnmount();
  }

  Route? get _currentRoute {
    final controller = cc.read(wlib.routerCapsule);

    if (!controller.initied) {
      return null;
    }

    return controller.getRoute().casted;
  }

  @override
  JsonMap get styleMap {
    final loading = _loading;

    return {
      ...super.styleMap,
      'position': 'relative',
      'height': '100%',
      'display': 'flex',
      'flexDirection': 'column',
      'overflow': 'hidden',
      if (loading) 'backgroundColor': 'rgba(10, 10, 10)', // Black
      if (!loading) 'backgroundColor': 'rgba(40, 40, 40)', // Grey
    };
  }
}
```

This is the `A` capsule:
```dart
LoadingModeController loadingModeCapsule(CapsuleHandle use) {
  final (loading, setLoading) = use.state(false);
  final log = use(loggerCapsule);

  return (
    loading: loading,
    toggle: () => setLoading(!loading),
    // on: () => setLoading(true),
    on: () {
      log.d('Setting loading mode ON -- current: $loading');
      setLoading(true);
    },
    // off: () => setLoading(false),
    off: () {
      log.d('Setting loading mode OFF -- current: $loading');
      setLoading(false);
    },
  );
}

typedef LoadingModeController = ({
  bool loading,
  void Function() toggle,
  void Function() on,
  void Function() off,
});
```


This is the output:
```
�[38;5;244m[T]�[0m  Websocket connection -- On done (Stream) -- 1707105384312707 [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
�[38;5;244m[T]�[0m  Websocket connection -- On done (Sink) -- 1707105384312707 -- Skipping due to yet done [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- websocket state listener -- WebsocketConnectionState.disconnected [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- SETTING LOADING ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Setting loading mode ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Connecting to Webosckets Server [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- websocket state listener -- WebsocketConnectionState.connecting [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Connecting to wss://pandemic-plandemic.ddns.net/ws [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
GET
wss://pandemic-plandemic.ddns.net/ws

Firefox no puede establecer una conexión con el servidor en wss://pandemic-plandemic.ddns.net/ws. [main.dart.js:5007:7](https://pandemic-plandemic.ddns.net/main.dart.js)
�[38;5;196m[E]�[0m  Error connecting to websocket server -- WebSocketChannelException: WebSocket connection failed. [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- websocket state listener -- WebsocketConnectionState.disconnected [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- SETTING LOADING ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Setting loading mode ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Trying again after 5 seconds delay [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Connecting to Webosckets Server [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- websocket state listener -- WebsocketConnectionState.connecting [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Connecting to wss://pandemic-plandemic.ddns.net/ws [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
GET
wss://pandemic-plandemic.ddns.net/ws

Firefox no puede establecer una conexión con el servidor en wss://pandemic-plandemic.ddns.net/ws. [main.dart.js:5007:7](https://pandemic-plandemic.ddns.net/main.dart.js)
�[38;5;196m[E]�[0m  Error connecting to websocket server -- WebSocketChannelException: WebSocket connection failed. [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- websocket state listener -- WebsocketConnectionState.disconnected [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[DEBUG] -- SETTING LOADING ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Setting loading mode ON -- current: false [main.dart.js:5044:78](https://pandemic-plandemic.ddns.net/main.dart.js)
[D]  Trying again after 5 seconds delay
```

As you can see in the output, it tries repeatedly to set loading on (`SETTING LOADING ON`) but always its value is false (is not getting updated).
Basically I'm turning down the server, so the weboskcet connection get closed and the loading spinner must be shown, but it isn't.

Sorry about this mess. I'm going to try to make a video clarifying this.

**Comment by busslina** [comment]
Video 1: https://mega.nz/file/IWwFCB7Q#ByvYFkgB5rKlpa5WtSTZtJoetTN5apHNgwnlYkaJfUc
Video 2:

**Comment by busslina** [comment]
Basically, having this `Setting loading mode ON -- current: false` seems meaning that something weird is happening, and it's only happening when executing that setState from inside the listener

**Comment by busslina** [comment]
I simulated this issue in this repo and works well... so I don't know what the heck its happening on my project. I will try to resolve it later.

https://github.com/busslina/rearch-listener-issue

**Comment by busslina** [comment]
Hi, I add here a little bit more info to the one that I shared with you via Discord:

I modfied the code of `lazyStateGetterSetter` side effect in order to know whats happening there:
```dart
(T Function(), void Function(T)) lazyStateGetterSetter<T>(
    T Function() init, {
    String? debugName,
  }) {
    void debug(String msg) {
      if (debugName == null) {
        return;
      }

      print('[$debugName]: $msg');
    }

    // We use register directly to keep the same setter function across rebuilds
    // (but we need to return a new getter on each build, see below for more)
    final (getter, setter) = use.register((api) {
      debug('lazyStateGetterSetter() -- 1 -- init');

      var state = init();

      T getter() => state;
      void setter(T newState) {
        debug('lazyStateGetterSetter() -- 2 -- setter');

        api.rebuild((cancelRebuild) {
          debug('lazyStateGetterSetter() -- 3 -- setter');
          if (newState == state) {
            debug('lazyStateGetterSetter() -- 4 -- setter');
            cancelRebuild();
            return;
          }

          debug('lazyStateGetterSetter() -- 5 -- setter');

          state = newState;

          debug('lazyStateGetterSetter() -- 6 -- setter -- new state: $state');
        });
      }

      return (getter, setter);
    });
```

Firstly this snapshot shows the predicted/non-error debug:
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/556ebe24-2c46-4ab8-8f29-3241c76bf1f5)

This is also good, when the state doesn't change because it is the same:
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/adce6ae8-fcf9-4943-be46-8a11132261a5)

Lastly, this snapshot shows what is being executed on the weird-error case:
![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/bfc49ffa-905c-43c1-9fb6-309de817db68)

As you can see, the setter is being executed, but no the inner `api.rebuild`callback.

**Comment by GregoryConrad** [comment]
> As you can see, the setter is being executed, but not the inner api.rebuildcallback.

That’d be because the value passed into the setter doesn’t change. In other words, the value returned by getValue at that instant in time is equal to the value you’re setting.

**Comment by GregoryConrad** [comment]
Oh wait, you’re saying the callback to api.rebuild isn’t being executed. I’ll take a look more into that

**Comment by busslina** [comment]
> > As you can see, the setter is being executed, but not the inner api.rebuildcallback.
> 
> That’d be because the value passed into the setter doesn’t change. In other words, the value returned by getValue at that instant in time is equal to the value you’re setting.

I think you are wrong here, because as I shown in my last message, you can see the second screenshot is when the value is the same and the state is not updated.

But you also can see the third screenshot. It don't even do the state comparison to check if it is the same state.

Also, I show in the first messages, it get stuck in one state (false) and doesn't want to change it (to true).

Again, my last message with the three snapshots show a bug in my opinion

**Comment by GregoryConrad** [comment]
Yup, caught that after I sent my first msg. Looking into it, might be getting back to you tomorrow depending on how things shake out

**Comment by busslina** [comment]
The state comparison is done inside the `api.rebuild`, but it doesn't get executed, so how can it know if it is a repeated state (which it isn't)?

**Comment by busslina** [comment]
> Yup, caught that after I sent my first msg. Looking into it, might be getting back to you tomorrow depending on how things shake out

Thank you

**Comment by GregoryConrad** [comment]
Ohh. Think I got it

```dart

if (state.disconnected) {
        debug('SETTING LOADING ON -- current: ${loadingMode.loading}');

        // Setting loading
        loadingMode.on();

```

loadingMode.on() sets the state, and this is run within a listener. _This is not allowed._ there is an open issue to add an assertion for this check: https://github.com/GregoryConrad/rearch-dart/issues/17
Sorry that assertion isn’t there yet, would’ve saved you a ton of headache.

as for quick fix: toss the .on() call in a `Future.delayed` or after an `await null;`
A proper fix would involve refactoring to prevent one state from needing to be updated when another capsules state is updated. In this case, loading state is being updated via the websocket state capsule, so it’s likely that these two states should be directly related in the same capsule instead of separated into different capsules. (The alternative would be to update their states together via txns, but it’s likely you just have state in the wrong spot right now)

**Comment by busslina** [comment]
Okay. I knew that the issue was related on being executed inside a listener, that's why the issue title that I choosed. Also I tested it in a simplwtest project and worked (to modify a state from a listener).

Okay, I will remember this restriction and the workaround you propose me.

Is it a perpetual restriction or just a current one that may be possible to remove in the future?

Thanks ;)

**Comment by busslina** [comment]
I will try the transaction workaround because I found both capsules to be different entities

**Comment by GregoryConrad** [comment]
> Okay. I knew that the issue was related on being executed inside a listener, that's why the issue title that I choosed.

Oops, didn't read it close enough haha 😅

> Also I tested it in a simple test project and worked (to modify a state from a listener).

Yea; I consider it UB. It may work, it may not. Thus why I want to add the assertion in so that it doesn't trip anyone up.

> Is it a perpetual restriction or just a current one that may be possible to remove in the future?

It's an intentional restriction, because it will allows for incorrect data modeling. I wrote some about proper/derived state in the documentation [here](https://rearch.gsconrad.com/core/effects#transactions).

In your situation, if it isn't possible for `loadingModeCapsule` to simply derive its state from `websocketStateCapsule` (why can't it?), then you would have to use transactions to update the state of both of those capsules at the same time. But again, on the surface  this seems like an instance where you should be using derived state instead of transactions.

**Comment by busslina** [comment]
> (why can't it?)
The thing is that both capsules are defined at a library level, but maybe I want to link them at project level.

But I take your words/advice seriously so I can follow the best guidelines. For now I did this approach:

```dart
extension _ConvenienceExtension on SideEffectRegistrar {
  SideEffectRegistrar get use => this;
}

extension SideEffectRegistrarExtension on SideEffectRegistrar {
  LoadingModeController loadingMode() {
    final (state, setState) = use.state<({bool loading, String? message})>(
        (loading: false, message: null));

    return (
      loading: state.loading,
      loadingMessage: state.message,
      toggle: ({String? message}) => setState(
          (loading: !state.loading, message: state.loading ? message : null)),
      on: ({String? message, bool asyncMode = false}) {
        // Async mode
        if (asyncMode) {
          Future.microtask(() => setState((loading: true, message: message)));
          return;
        }

        setState((loading: true, message: message));
      },
      off: ({bool asyncMode = false}) {
        // Async mode
        if (asyncMode) {
          Future.microtask(() => setState((loading: false, message: null)));
          return;
        }

        setState((loading: false, message: null));
      },
    );
  }
}

typedef LoadingModeController = ({
  bool loading,
  String? loadingMessage,
  void Function({String? message}) toggle,
  void Function({String? message, bool asyncMode}) on,
  void Function({bool asyncMode}) off,
});
```

You can see `asyncMode`

**Comment by busslina** [comment]
BTW, I'm thinking on create a dropdown list button for showing the "Developed with" or "Powered by" technologies I use: Dart, React, ... Have you any inconvenient on me putting ReArch on the list? Maybe you don't want it in a controversial site like this:

![imagen](https://github.com/GregoryConrad/rearch-dart/assets/73592852/7d27b5bc-5229-4c06-8279-b3eeae415255)

**Comment by GregoryConrad** [comment]
> Have you any inconvenient on me putting ReArch on the list? Maybe you don't want it in a controversial site like this:

You're welcome to list it. Just don't say I or my projects endorse that particular project hahaha 😆

**Comment by busslina** [comment]
> You're welcome to list it. Just don't say I (or my projects) endorse that particular project hahaha 😆

Of course, done xD
